src/lib/optimizer/strategy/like_replacement_rule.cpp

```diff
@@ -0,0 +1,73 @@
+#include "like_replacement_rule.hpp"
+
+#include <algorithm>
+#include <iostream>
+
+#include "all_parameter_variant.hpp"
+#include "constant_mappings.hpp"
+#include "expression/expression_functional.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+#include "operators/operator_scan_predicate.hpp"
+#include "statistics/chunk_statistics/chunk_statistics.hpp"
+#include "storage/storage_manager.hpp"
+#include "storage/table.hpp"
+#include "utils/assert.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+std::string LikeReplacementRule::name() const { return "Like Replacement Rule"; }
+
+void LikeReplacementRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
+  const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(node);
+  if (predicate_node != nullptr) {
+    const auto expression = std::dynamic_pointer_cast<BinaryPredicateExpression>(predicate_node->predicate());
+    if (expression != nullptr && expression->predicate_condition == PredicateCondition::Like) {
+      const auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression->left_operand());
+      const auto value_expression = std::dynamic_pointer_cast<ValueExpression>(expression->right_operand());
+      if (value_expression != nullptr && column_expression != nullptr) {
+        const auto value = boost::get<std::string>(value_expression->value);
+        const auto offset = value.find_first_of("%");
+        // Check if the string has an % wildcard and ends with the % wildcard
+        if (offset != std::string::npos && offset > 0 && value.length() == offset + 1) {
+          // Calculate lower and upper bound of the string
+          const auto lower_bound = value.substr(0, offset);
+          const auto current_character_value = static_cast<int>(lower_bound.at(lower_bound.length() - 1));
+          // Find next value according to ASCII-table
+          if (current_character_value > 0 && current_character_value < 127) {
+            const auto next_character = static_cast<char>(current_character_value + 1);
```

Comment:
You stated in your description that it works for ASCII only. Maybe you could outsource this line to a function that "knows" about the charset and how to get the next character? It might not make much sense for ascii only, but e.g. if we consider supporting UTF-8 with its var-length characters, the next-char-function might get more complicated.
I feel like this could be easier to extend

Reply:
I do not know how Hyrise deals with collations (especially case-(in)sensitivity), so maybe the comment below is not yet relevant, and maybe will never be.

Anyway, consider the following scenario: we get a query containing `name LIKE "Z%"`.
Your rule will increment `Z` (90) to `[` (91). While this might look a bit weird, it should probably still work fine, as long as strings are compared charwise by their ascii code points (no idea if this is the case for all collations?).
Further, there could be a problem when using case-insensitive comparisons. "Could" because it depends on the concrete implementation, but assume we achieve case-insensitivity by e.g. lowercasing everything, and we do this when the LQP is done with optimization.
In that case, the `Z` (90) will become `z` (122), while `[` stays at 91. After your rule it would look like `name >= 122 AND name < 91`, which will never evaluate to true.

Again, I know there's a lot of ifs in there, so I am not sure if this really needs fixing. If it needs a fix, one idea could be to make the condition in L. 41 stricter, e.g. only apply the rule if the last char is in A-Y or a-y.